execution/executor-benchmark/src/account_generator.rs

```diff
@@ -78,42 +93,89 @@ impl AccountCache {
+            .map(|a| Arc::try_unwrap(a).unwrap().into_inner())
+            .collect();
+        let other: Vec<LocalAccount> = other
+            .into_iter()
+            .map(|a| Arc::try_unwrap(a).unwrap().into_inner())
+            .collect();
+        (accounts, other)
     }
 
     pub fn len(&self) -> usize {
         self.accounts.len()
     }
 
-    pub fn accounts(&self) -> &VecDeque<LocalAccount> {
+    pub fn accounts(&self) -> &VecDeque<Arc<RwLock<LocalAccount>>> {
         &self.accounts
     }
 
     pub fn grow(&mut self, n: usize) {
-        let accounts: Vec<_> = (0..n).map(|_| self.generator.generate()).collect();
+        let accounts: Vec<_> = (0..n)
+            .map(|_| Arc::new(RwLock::new(self.generator.generate())))
+            .collect();
         self.accounts.extend(accounts);
     }
+}
 
-    pub fn get_random(&mut self) -> &mut LocalAccount {
-        let indices = rand::seq::index::sample(&mut self.rng, self.accounts.len(), 1);
-        let index = indices.index(0);
-
-        &mut self.accounts[index]
-    }
-
-    pub fn get_random_transfer_batch(
+impl RandomAccountGenerator for &mut AccountCache {
+    fn get_random_transfer_batch(
         &mut self,
         batch_size: usize,
-    ) -> (&mut LocalAccount, Vec<AccountAddress>) {
+    ) -> (Arc<RwLock<LocalAccount>>, Vec<AccountAddress>) {
         let indices = rand::seq::index::sample(&mut self.rng, self.accounts.len(), batch_size + 1);
         let sender_idx = indices.index(0);
         let receivers = indices
             .iter()
             .skip(1)
-            .map(|i| self.accounts[i].address())
+            .map(|i| self.accounts[i].read().address())
             .collect();
-        let sender = &mut self.accounts[sender_idx];
+        let sender = self.accounts[sender_idx].clone();
 
         (sender, receivers)
     }
+
+    fn get_random(&mut self) -> Arc<RwLock<LocalAccount>> {
+        let indices = rand::seq::index::sample(&mut self.rng, self.accounts.len(), 1);
+        let index = indices.index(0);
+
+        self.accounts[index].clone()
+    }
+}
+
+pub struct NoConflictsAccountCache {
+    pub accounts: VecDeque<Arc<RwLock<LocalAccount>>>,
+}
+
+impl NoConflictsAccountCache {
+    const SEED: Seed = [1; 32];
+
+    pub fn new(account_cache: &AccountCache) -> Self {
+        let mut rng = StdRng::from_seed(Self::SEED);
+        let mut accounts = account_cache.accounts.clone();
+        accounts.make_contiguous().shuffle(&mut rng);
+        Self { accounts }
+    }
+}
+
+impl RandomAccountGenerator for NoConflictsAccountCache {
+    fn get_random(&mut self) -> Arc<RwLock<LocalAccount>> {
+        // Since the accounts are already shuffled, we can just pop the first one.
+        self.accounts.pop_front().unwrap()
+    }
+
+    fn get_random_transfer_batch(
+        &mut self,
+        batch_size: usize,
+    ) -> (Arc<RwLock<LocalAccount>>, Vec<AccountAddress>) {
+        let sender = self.accounts.pop_front().unwrap();
```

Comment:
So to avoid panic, it's caller's responsibility to make sure it calls the methods correct number of times, right, as NoConflictsAccountCache pops the accounts. I think our use-cases seem fine, but in general seems easy to get into trouble. Returning a Result would be worse because AccountCache doesn't have this problem.
We could add size method or smt separate to check these cases, but maybe it's okay for now since this is for benchmarks. I'd just propose to use expect("not enough accounts") or some error message on all the pop_front. unwrap()'s

Reply:
Good point, I added a size method and added an explicit check on the caller side.